src/basedtyping/transformer.py

```diff
@@ -0,0 +1,231 @@
+        elif localns is None:
+            # apparently pyright doesn't infer this automatically
+            assert globalns is not None
+            localns = globalns
+
+        self.typing_name = f"typing_extensions_{uuid.uuid4().hex}"
+        self.basedtyping_name = f"basedtyping_{uuid.uuid4().hex}"
+        self.globalns = globalns
+        import typing_extensions
+
+        self.localns = {
+            **localns,
+            self.typing_name: typing_extensions,
+            self.basedtyping_name: basedtyping,
+        }
+
+    def visit(self, node: ast.AST) -> ast.AST:
+        return cast(ast.AST, super().visit(node))
+
+    def eval_type(
+        self,
+        node: ast.FunctionType | ast.Expression | ast.expr,
+        *,
+        original_ref: typing.ForwardRef | None = None,
+    ) -> object:
+        if isinstance(node, ast.expr):
+            node = ast.copy_location(ast.Expression(node), node)
+        ref = typing.ForwardRef(ast.dump(node))
+        if original_ref:
+            for attr in ("is_argument", " is_class", "module"):
+                attr = f"__forward_{attr}__"
+                if hasattr(original_ref, attr):
+                    setattr(ref, attr, cast(object, getattr(original_ref, attr)))
+        if not isinstance(node, ast.FunctionType):
+            ref.__forward_code__ = compile(node, "<node>", "eval")
+        try:
+            type_ = typing._type_convert(  # type: ignore[attr-defined]
+                cast(object, eval(ref.__forward_code__, self.globalns, self.localns))  # noqa: S307
+            )
+            return typing._eval_type(  # type: ignore[attr-defined]
+                type_, self.globalns, self.localns
+            )
+        except TypeError:
+            return None
+
+    def _typing(self, attr: str) -> ast.Attribute:
+        result = ast.Attribute(
+            value=ast.Name(id=self.typing_name, ctx=ast.Load()), attr=attr, ctx=ast.Load()
+        )
+        return ast.fix_missing_locations(result)
+
+    def _basedtyping(self, attr: str) -> ast.Attribute:
+        result = ast.Attribute(
+            value=ast.Name(id=self.basedtyping_name, ctx=ast.Load()), attr=attr, ctx=ast.Load()
+        )
+        return ast.fix_missing_locations(result)
+
+    def _literal(self, value: ast.Constant | ast.Name | ast.Attribute) -> ast.Subscript:
+        return self.subscript(self._typing("Literal"), value)
+
+    def subscript(self, value: ast.expr, slice_: ast.expr) -> ast.Subscript:
+        result = ast.Subscript(value=value, slice=slice_, ctx=ast.Load())
+        return ast.fix_missing_locations(result)
+
+    _implicit_tuple = False
+
+    @contextmanager
+    def implicit_tuple(self) -> typing.Iterator[None]:
+        implicit_tuple = self._implicit_tuple
+        self._implicit_tuple = True
+        try:
+            yield
+        finally:
+            self._implicit_tuple = implicit_tuple
+
+    def visit_Subscript(self, node: ast.Subscript) -> ast.AST:
```

Comment:
should these have `@override`? if so why didnt mypy complain

Reply:
minimum version moments

Reply:
what that means